### PR TITLE
converter: support ref image build

### PR DESCRIPTION
**Pull request by imeoer**
It allows nydus lazily load OCI targz layer with the nydus zran ref image.

Related nydus components implementation: https://github.com/dragonflyoss/image-service/pull/858

### CODE REVIEW

File: pkg/ref/ref.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+		return fmt.Errorf("can't find ref and digest from label %+v", labels)
+	}
+
+	rc, err := m.resolver.Resolve(ref, layerDigest, labels)
+	if err != nil {
+		return errors.Wrapf(err, "resolve from ref %s, digest %s", ref, layerDigest)
+	}
+	defer rc.Close()
+
+	blobFile, err := os.CreateTemp(m.blobCacheDir, "downloading-")
```
Comment (changweige) [inline pkg/ref/ref.go:49]: I hope snapshotter does not deal with any layer downloads, which I think can be delagated to nydusd

**Comment by imeoer** [comment]
cc @changweige @jiangliu Waiting nydus-image/nydusd v2.2-rc1 to be released.

File: pkg/converter/tool/builder.go [hunk 2]
```diff
@@ -79,6 +88,8 @@ func Pack(option PackOption) error {
 		"--fs-version",
 		option.FsVersion,
 		"--inline-bootstrap",
+		// "--features",
+		// "blob_toc",
```
Comment (changweige) [inline pkg/converter/tool/builder.go:92]: Please avoid dead code if they are node used.

File: pkg/converter/tool/builder.go [hunk 3]
```diff
@@ -117,6 +128,46 @@ func Pack(option PackOption) error {
...
+	args = append(args, option.SourcePath)
+
+	ctx := context.Background()
+	var cancel context.CancelFunc
+	if option.Timeout != nil {
+		ctx, cancel = context.WithTimeout(ctx, *option.Timeout)
+		defer cancel()
+	}
+
+	logrus.Debugf("\tCommand: %s %s", option.BuilderPath, strings.Join(args, " "))
```
Comment (changweige) [inline pkg/converter/tool/builder.go:153]: Should avoid print messages in a library package?

File: pkg/converter/tool/builder.go [hunk 4]
```diff
@@ -117,6 +128,46 @@ func Pack(option PackOption) error {
...
+
+	cmd := exec.CommandContext(ctx, option.BuilderPath, args...)
+	cmd.Stdout = logger.Writer()
+	cmd.Stderr = logger.Writer()
+
+	if err := cmd.Run(); err != nil {
+		if errdefs.IsSignalKilled(err) && option.Timeout != nil {
+			logrus.WithError(err).Errorf("fail to run %v %+v, possibly due to timeout %v", option.BuilderPath, args, *option.Timeout)
```
Comment (changweige) [inline pkg/converter/tool/builder.go:161]: Ditto

File: pkg/converter/types.go [hunk 5]
```diff
@@ -48,6 +64,8 @@ type PackOption struct {
 	PrefetchPatterns string
 	// Compressor specifies nydus blob compression algorithm.
 	Compressor string
+	// OCIRef enables to convert OCI tar(.gz) blob to nydus referenced blob.
```
Comment (changweige) [inline pkg/converter/types.go:67]: enables converting OCI to nydus

File: pkg/converter/types.go [hunk 6]
```diff
@@ -82,3 +102,58 @@ type UnpackOption struct {
 	// Timeout cancels execution once exceed the specified time.
 	Timeout *time.Duration
 }
+
+type TOCEntry struct {
+	/// Feature flags of entry
```
Comment (changweige) [inline pkg/converter/types.go:107]: Does comment style `///` has special use case? It's Rust style ....

**Comment by changweige** [comment]
What tool is relying on this PR? nerdctl?

**Comment by imeoer** [comment]
> What tool is relying on this PR? nerdctl?

Yes, acceld or nerdctl.

File: pkg/converter/tool/builder.go (hunk 3 above)
Comment (imeoer) [reply pkg/converter/tool/builder.go:153]: It's ok for printing debug messages.

**Comment by changweige** [comment]
I am willing to experience the conversion added by this PR. What should I do?

**Comment by imeoer** [comment]
@changweige You can follow the doc https://github.com/dragonflyoss/image-service/blob/master/docs/nydus-zran.md#generate-nydus-zran-artifact

File: pkg/converter/types.go (hunk 6 above)
Comment (imeoer) [reply pkg/converter/types.go]: Fixed

File: pkg/converter/tool/builder.go (hunk 2 above)
Comment (imeoer) [reply pkg/converter/tool/builder.go]: Fixed

**Pull request merged**